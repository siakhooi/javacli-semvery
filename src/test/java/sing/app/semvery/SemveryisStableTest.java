package sing.app.semvery;

import static com.github.stefanbirkner.systemlambda.SystemLambda.tapSystemOut;
import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static sing.app.semvery.TestData.createArguments;
import java.util.ArrayList;
import java.util.stream.Stream;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;

@ExtendWith({SnapshotExtension.class})
class SemveryisStableTest {

    private Semvery app;
    private Expect expect;

    @BeforeEach
    void setup() {
        this.app = new Semvery();
    }

    @ParameterizedTest
    @MethodSource
    void testIsStable(String scenario, ReturnValue returnValue, String[] arguments)
            throws Exception {
        String text = tapSystemOut(() -> assertEquals(returnValue, app.run(arguments)));
        assertDoesNotThrow(() -> expect.scenario(scenario).toMatchSnapshot(text));
    }

    private static Stream<Arguments> testIsStable() {
        ArrayList<Arguments> a = new ArrayList<>();

        final ReturnValue ok = ReturnValue.OK;
        final ReturnValue nok = ReturnValue.NOT_OK;
        final String OPERATOR = "isStable";
        final String[] ANY = new String[] {"--any"};
        final String[] EMPTY = new String[] {};

        for (String o : TestData.OPERATIONS) {
            a.addAll(createArguments("all-stable", ok, o, OPERATOR, EMPTY, TestData.ALL_STABLE));
            a.addAll(createArguments("all-stable", ok, o, OPERATOR, ANY, TestData.ALL_STABLE));

            a.addAll(createArguments("all-not-valid", nok, o, OPERATOR, EMPTY,
                    TestData.ALL_INVALID));
            a.addAll(createArguments("all-not-valid", nok, o, OPERATOR, ANY, TestData.ALL_INVALID));

            a.addAll(createArguments("all-not-stable", nok, o, OPERATOR, EMPTY,
                    TestData.ALL_NOT_STABLE));
            a.addAll(createArguments("all-not-stable", nok, o, OPERATOR, ANY,
                    TestData.ALL_NOT_STABLE));


            a.addAll(createArguments("mixed-stable-not-valid", nok, o, OPERATOR, EMPTY,
                    TestData.STABLE_INVALID));
            a.addAll(createArguments("mixed-stable-not-valid", ok, o, OPERATOR, ANY,
                    TestData.STABLE_INVALID));

            a.addAll(createArguments("mixed-stable-not-stable", nok, o, OPERATOR, EMPTY,
                    TestData.STABLE_NOT_STABLE));
            a.addAll(createArguments("mixed-stable-not-stable", ok, o, OPERATOR, ANY,
                    TestData.STABLE_NOT_STABLE));

            a.addAll(createArguments("mixed-stable-not-stable-not-valid", nok, o, OPERATOR, EMPTY,
                    TestData.STABLE_NOT_STABLE_NOT_VALID));
            a.addAll(createArguments("mixed-stable-not-stable-not-valid", ok, o, OPERATOR, ANY,
                    TestData.STABLE_NOT_STABLE_NOT_VALID));
        }

        return a.stream();
    }
}
