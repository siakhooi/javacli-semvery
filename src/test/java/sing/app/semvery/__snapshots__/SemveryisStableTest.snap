sing.app.semvery.SemveryisStableTest.testIsStable[all-not-stable:0.1.0,1.2.3-4,1.2.3-BETA,33.5454.54353-pre]=[
Value                Result              
-----                -----               
0.1.0                not stable          
1.2.3-4              not stable          
1.2.3-BETA           not stable          
33.5454.54353-pre    not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-stable:0.2.0]=[
Value                Result              
-----                -----               
0.2.0                not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-stable:1.2.3-4]=[
Value                Result              
-----                -----               
1.2.3-4              not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-stable:1.2.3-BETA]=[
Value                Result              
-----                -----               
1.2.3-BETA           not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-stable:33.5454.54353-pre]=[
Value                Result              
-----                -----               
33.5454.54353-pre    not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-valid:1.0.A]=[
Value                Result              
-----                -----               
1.0.A                invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-valid:1.0.B,adfadfasjdfsd,23423,2423.2342342,24234.234234.]=[
Value                Result              
-----                -----               
1.0.B                invalid             
adfadfasjdfsd        invalid             
23423                invalid             
2423.2342342         invalid             
24234.234234.        invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-valid:23423]=[
Value                Result              
-----                -----               
23423                invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-valid:2423.2342342]=[
Value                Result              
-----                -----               
2423.2342342         invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-valid:24234.234234.]=[
Value                Result              
-----                -----               
24234.234234.        invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-not-valid:adfadfasjdfsd]=[
Value                Result              
-----                -----               
adfadfasjdfsd        invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-stable:1.0.0]=[
Value                Result              
-----                -----               
1.0.0                stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-stable:1.2.3+sHa.0nSFGKjk]=[
Value                Result              
-----                -----               
1.2.3+sHa.0nSFGKjk   stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-stable:1.2.3,24434.24233.23433,1.2.3+sHa.0nSFGjkf]=[
Value                Result              
-----                -----               
1.2.3                stable              
24434.24233.23433    stable              
1.2.3+sHa.0nSFGjkf   stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[all-stable:23434.24343.2343]=[
Value                Result              
-----                -----               
23434.24343.2343     stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-stable-not-valid:0.1.0,1.1.A,3.0.0]=[
Value                Result              
-----                -----               
0.1.0                not stable          
1.1.A                invalid             
3.0.0                stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-stable-not-valid:1.1.0,3.0.A,0.1.0]=[
Value                Result              
-----                -----               
1.1.0                stable              
3.0.A                invalid             
0.1.0                not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-stable-not-valid:1.1.A,0.1.0,3.0.0]=[
Value                Result              
-----                -----               
1.1.A                invalid             
0.1.0                not stable          
3.0.0                stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-stable:0.1.0,1.1.0,3.0.0]=[
Value                Result              
-----                -----               
0.1.0                not stable          
1.1.0                stable              
3.0.0                stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-stable:1.1.0,0.1.0,3.0.0]=[
Value                Result              
-----                -----               
1.1.0                stable              
0.1.0                not stable          
3.0.0                stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-stable:1.1.0,3.0.0,0.1.0]=[
Value                Result              
-----                -----               
1.1.0                stable              
3.0.0                stable              
0.1.0                not stable          

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-valid:1.0.0,1.1.0,3.0.A]=[
Value                Result              
-----                -----               
1.0.0                stable              
1.1.0                stable              
3.0.A                invalid             

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-valid:1.0.A,1.1.0,1.3.0]=[
Value                Result              
-----                -----               
1.0.A                invalid             
1.1.0                stable              
1.3.0                stable              

]


sing.app.semvery.SemveryisStableTest.testIsStable[mixed-stable-not-valid:1.3.0,1.1.A,1.3.0]=[
Value                Result              
-----                -----               
1.3.0                stable              
1.1.A                invalid             
1.3.0                stable              

]